builtin/mainmenu/tab_online.lua

```diff
@@ -210,21 +210,66 @@ local function search_server_list(input)
 	for i = 1, #serverlistmgr.servers do
 		local server = serverlistmgr.servers[i]
 		local found = 0
+		local sername_keywords = {}
+		local sername_keywords_matchtimes = {}
+		local server_desc_keywords = {}
+		local server_desc_keywords_matchtimes = {}
+		-- Make search precise for exactly matched server
+		-- Or at least more exactly matched server should be rated higher at the match scoring
+		if server.name then
+			for word in server.name:gmatch("%S+") do
+				word = word:gsub("(%W)", "%%%1"):lower()
+				table.insert(sername_keywords, word)
+				if sername_keywords_matchtimes[word] == nil then
+					sername_keywords_matchtimes[word] = 1
+				else
+					sername_keywords_matchtimes[word] = sername_keywords_matchtimes[word] + 1
+				end
+			end
+		end
+		if server.description then
+			for word in server.description:gmatch("%S+") do
+				word = word:gsub("(%W)", "%%%1"):lower()
+				table.insert(server_desc_keywords, word)
+				if server_desc_keywords_matchtimes[word] == nil then
+					server_desc_keywords_matchtimes[word] = 1
+				else
+					server_desc_keywords_matchtimes[word] = server_desc_keywords_matchtimes[word] + 1
+				end
+			end
+		end
 		for k = 1, #keywords do
-			local keyword = keywords[k]
-			if server.name then
-				local sername = server.name:lower()
-				local _, count = sername:gsub(keyword, keyword)
-				found = found + count * 4
+			local keyword
```

Comment:
This entire loop body is also duplicated (but interspersed) code in a trenchcoat. You're effectively running the same logic for the description and the server name, just with different weights. This should again be a function.